### ISSUE TITLE
Supervised clustering and forest embeddings

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10804]
Supervised clustering and forest embeddings<br>
<br>
https://ift.tt/nBRwZ9d<br>
<br>
Guilherme Duarte Marmerola<br>
<br>

<div><div><div><div><article><p>In this post, I’ll try out a new way to represent data and perform clustering: forest embeddings. A forest embedding is a way to represent a feature space using a random forest. Each data point $x_i$ is encoded as a vector $x_i = [e_0, e_1, …, e_k]$ where each element $e_i$ holds which leaf of tree $i$ in the forest $x_i$ ended up into. The encoding can be learned in a supervised or unsupervised manner:</p>
<ol>
<li><p><strong>Supervised:</strong> we train a forest to solve a regression or classification problem. Then, we use the trees structure to extract the embedding.</p></li>
<li><p><strong>Unsupervised:</strong> each tree of the forest builds splits at random, without using a target variable.</p></li>
</ol>
<p>There may be a number of benefits in using forest-based embeddings:</p>
<ol>
<li><p><strong>Distance calculations are ok when there are categorical variables:</strong> as we’re using leaf co-ocurrence as our similarity, we do not need to be concerned that distance is not defined for categorical variables.</p></li>
<li><p><strong>For supervised embeddings, we automatically set optimal weights for each feature for clustering:</strong> if we want to cluster our data given a target variable, our embedding automatically selects the most relevant features.</p></li>
<li><p><strong>We do not need to worry about scaling features:</strong> we do not need to worry about the scaling of the features, as we’re using decision trees.</p></li>
</ol>
<p>In the next sections, we implement some simple models and test cases. You can find the complete code at my <a href="https://github.com/gdmarmerola/forest-embeddings">GitHub page</a>.</p>
<h2>Building the embeddings</h2>
<p>So how do we build a forest embedding? It’s very simple. We start by choosing a model. In our case, we’ll choose any from <code>RandomTreesEmbedding</code>, <code>RandomForestClassifier</code> and <code>ExtraTreesClassifier</code> from <code>sklearn</code>. Let’s say we choose <code>ExtraTreesClassifier</code>. The first thing we do, is to fit the model to the data.</p>
<div><div><pre><code><span># choosing and fitting a model
</span><span>model</span> <span>=</span> <span>ExtraTreesClassifier</span><span>(</span><span>n_estimators</span><span>=</span><span>100</span><span>,</span> <span>min_samples_leaf</span><span>=</span><span>10</span><span>)</span>
<span>model</span><span>.</span><span>fit</span><span>(</span><span>X</span><span>,</span> <span>y</span><span>)</span>
</code></pre></div></div>
<p>As we’re using a supervised model, we’re going to learn a supervised embedding, that is, the embedding will weight the features according to what is most relevant to the target variable. Data points will be closer if they’re similar in the most relevant features. After model adjustment, we apply it to each sample in the dataset to check which leaf it was assigned to.</p>
<div><div><pre><code><span># let us apply to X
</span><span>leaves</span> <span>=</span> <span>model</span><span>.</span><span>apply</span><span>(</span><span>X</span><span>)</span>
</code></pre></div></div>
<p>Then, we apply a sparse one-hot encoding to the <code>leaves</code>:</p>
<div><div><pre><code><span># applying a one-hot encoding scheme
</span><span>M</span> <span>=</span> <span>OneHotEncoder</span><span>().</span><span>fit_transform</span><span>(</span><span>leaves</span><span>)</span>
</code></pre></div></div>
<p>At this point, we could use an efficient data structure such as a KD-Tree to query for the nearest neighbours of each point. To simplify, we use brute force and calculate all the pairwise co-ocurrences in the leaves using dot products:</p>
<div><div><pre><code><span># we perform M*M.transpose(), which is the same to # computing all the pairwise co-ocurrences in the leaves </span><span>S</span> <span>=</span> <span>(</span><span>M</span><span>*</span><span>M</span><span>.</span><span>transpose</span><span>()).</span><span>todense</span><span>()</span>

<span># lastly, we normalize and subtract from 1, to get dissimilarities
</span><span>D</span> <span>=</span> <span>1</span> <span>-</span> <span>S</span><span>/</span><span>S</span><span>.</span><span>max</span><span>()</span>
</code></pre></div></div>
<p>Finally, we have a <code>D</code> matrix, which counts how many times two data points have <strong>not</strong> co-occurred in the tree leaves, normalized to the [0,1] interval. <code>D</code> is, in essence, a dissimilarity matrix.</p>
<p>The last step we perform aims to make the embedding easy to visualize. We feed our dissimilarity matrix <code>D</code> into the t-SNE algorithm, which produces a 2D plot of the embedding.</p>
<div><div><pre><code><span># computing 2D embedding with tsne, for visualization purposes
</span> <span>embed</span> <span>=</span> <span>TSNE</span><span>(</span><span>metric</span><span>=</span><span>'precomputed'</span><span>,</span> <span>perplexity</span><span>=</span><span>30</span><span>).</span><span>fit_transform</span><span>(</span><span>D</span><span>)</span>
</code></pre></div></div>
<p>In the next sections, we’ll run this pipeline for various toy problems, observing the differences between an unsupervised embedding (with <code>RandomTreesEmbedding</code>) and supervised embeddings (Ranfom Forests and Extremely Randomized Trees). We favor supervised methods, as <strong>we’re aiming to recover only the structure that matters to the problem, with respect to its target variable.</strong></p>
<h2>Two blobs, two dimensions</h2>
<p>Let us start with a dataset of two blobs in two dimensions.</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-1.png"></p>
<p>As the blobs are separated and there’s no noisy variables, we can expect that unsupervised and supervised methods can easily reconstruct the data’s structure thorugh our similarity pipeline. After we fit our three contestants (<code>RandomTreesEmbedding</code>, <code>RandomForestClassifier</code> and <code>ExtraTreesClassifier</code>) to the data, we can take a look at the similarities they learned and the plot below:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-2.png"></p>
<p>The red dot is our “pivot”, such that we show the similarity of all the points in the plot to the pivot in shades of gray, black being the most similar. Each plot shows the similarities produced by one of the three methods we chose to explore. Similarities by the RF are pretty much binary: points in the same cluster have 100% similarity to one another as opposed to points in different clusters which have zero similarity. ET and RTE seem to produce “softer” similarities, such that the pivot has at least some similarity with points in the other cluster.</p>
<p>Finally, let us check the t-SNE plot for our methods. In the upper-left corner, we have the actual data distribution, our ground-truth. The other plots show t-SNE reconstructions from the dissimilarity matrices produced by methods under trial.</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-3.png"></p>
<p>All the embeddings give a reasonable reconstruction of the data, except for some artifacts on the ET reconstruction. I’m not sure what exactly are the artifacts in the ET plot, but they may as well be the t-SNE “overfitting” the local structure, close to the artificial clusters shown in the gaussian noise example in <a href="https://distill.pub/2016/misread-tsne/">here</a>.</p>
<h2>Two moons, two dimensions</h2>
<p>Now, let us check a dataset of two moons in two dimensions, like the following:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-4.png"></p>
<p>The similarity plot shows some interesting features:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-5.png"></p>
<p>And the t-SNE plot shows some weird patterns for RF and good reconstruction for the other methods:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-6.png"></p>
<p>RTE perfectly reconstucts the moon pattern, while ET unwraps the moons and RF shows a pretty strange plot. I think the ball-like shapes in the RF plot may correspond to regions in the space in which the samples could be perfectly classified in just one split, like, say, all the points in <strong>$y_1 &lt; -0.25$</strong>. All of these points would have 100% pairwise similarity to one another. As ET draws splits less greedily, similarities are softer and we see a space that has a more uniform distribution of points.</p>
<h2>Four moons, where only two are relevant</h2>
<p>Now, let us concatenate two datasets of moons, but we will only use the target variable of one of them, to simulate two irrelevant variables. The following plot shows the distribution for the four independent features of the dataset, $x_1$, $x_2$, $x_3$ and $x_4$. $x_1$ and $x_2$ are highly discriminative in terms of the target variable, while $x_3$ and $x_4$ are not. The following plot makes a good illustration:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-7.png"></p>
<p>The ideal embedding should throw away the irrelevant variables and reconstruct the true clusters formed by $x_1$ and $x_2$. Intuition tells us the only the supervised models can do this. As it’s difficult to inspect similarities in 4D space, we jump directly to the t-SNE plot:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-9.png"></p>
<p>As expected, supervised models outperform the unsupervised model in this case. RTE suffers with the noisy dimensions and shows a meaningless embedding. RF, with its binary-like similarities, shows artificial clusters, although it shows good classification performance. ET wins this competition showing only two clusters and slightly outperforming RF in CV.</p>
<h2>Real case: Boston housing dataset</h2>
<p>Finally, let us now test our models out with a real dataset: the <a href="http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_boston.html">Boston Housing dataset</a>, from the UCI repository. Here’s a snippet of it:</p>
<table>
<thead><tr>
<th>CRIM</th>
<th>ZN</th>
<th>INDUS</th>
<th>CHAS</th>
<th>NOX</th>
<th>RM</th>
<th>AGE</th>
<th>DIS</th>
<th>RAD</th>
<th>TAX</th>
<th>PTRATIO</th>
<th>B</th>
<th>LSTAT</th>
</tr></thead>
<tbody>
<tr>
<td>0.00632</td>
<td>18.0</td>
<td>2.31</td>
<td>0.0</td>
<td>0.538</td>
<td>6.575</td>
<td>65.2</td>
<td>4.0900</td>
<td>1.0</td>
<td>296.0</td>
<td>15.3</td>
<td>396.90</td>
<td>4.98</td>
</tr>
<tr>
<td>0.02731</td>
<td>0.0</td>
<td>7.07</td>
<td>0.0</td>
<td>0.469</td>
<td>6.421</td>
<td>78.9</td>
<td>4.9671</td>
<td>2.0</td>
<td>242.0</td>
<td>17.8</td>
<td>396.90</td>
<td>9.14</td>
</tr>
<tr>
<td>0.02729</td>
<td>0.0</td>
<td>7.07</td>
<td>0.0</td>
<td>0.469</td>
<td>7.185</td>
<td>61.1</td>
<td>4.9671</td>
<td>2.0</td>
<td>242.0</td>
<td>17.8</td>
<td>392.83</td>
<td>4.03</td>
</tr>
<tr>
<td>0.03237</td>
<td>0.0</td>
<td>2.18</td>
<td>0.0</td>
<td>0.458</td>
<td>6.998</td>
<td>45.8</td>
<td>6.0622</td>
<td>3.0</td>
<td>222.0</td>
<td>18.7</td>
<td>394.63</td>
<td>2.94</td>
</tr>
<tr>
<td>0.06905</td>
<td>0.0</td>
<td>2.18</td>
<td>0.0</td>
<td>0.458</td>
<td>7.147</td>
<td>54.2</td>
<td>6.0622</td>
<td>3.0</td>
<td>222.0</td>
<td>18.7</td>
<td>396.90</td>
<td>5.33</td>
</tr>
</tbody>
</table>
<p>This is a regression problem where the two most relevant variables are <code>RM</code> and <code>LSTAT</code>, accounting together for over 90% of total importance. We plot the distribution of these two variables as our reference plot for our forest embeddings. The color of each point indicates the value of the target variable, where yellow is higher. Let us check the t-SNE plot for our reconstruction methodologies.</p>
<p><img src="https://gdmarmerola.github.io//assets/img/forest_embeddings/emdeddings-img-10.png"></p>
<p>The first plot, showing the distribution of the most important variables, shows a pretty nice structure which can help us interpret the results. RTE is interested in reconstructing the data’s distribution, so it does not try to put points closer with respect to their value in the target variable. The supervised methods do a better job in producing a uniform scatterplot with respect to the target variable. Considering the two most important variables (90% gain) plot, ET is the closest reconstruction, while RF seems to have created artificial clusters. This is further evidence that ET produces embeddings that are more faithful to the original data distribution.</p>
<h2>Conclusion</h2>
<p>In this tutorial, we compared three different methods for creating forest-based embeddings of data. The unsupervised method Random Trees Embedding (RTE) showed nice reconstruction results in the first two cases, where no irrelevant variables were present. When we added noise to the problem, supervised methods could move it aside and reasonably reconstruct the real clusters that correlate with the target variable. Despite good CV performance, Random Forest embeddings showed instability, as similarities are a bit binary-like. However, Extremely Randomized Trees provided more stable similarity measures, showing reconstructions closer to the reality. We conclude that ET is the way to go for reconstructing supervised forest-based embeddings in the future.</p></article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Guilherme’s Blog <br>
<br>
February 5, 2024 at 04:52PM
